crypto/provider_core.c

```diff
@@ -1612,50 +1618,257 @@ OSSL_LIB_CTX *ossl_provider_libctx(const OSSL_PROVIDER *prov)
     return prov != NULL ? prov->libctx : NULL;
 }
 
-/* Wrappers around calls to the provider */
+/**
+ * @brief Prints the parameter values of an OSSL_PARAM array to a BIO.
+ *
+ * This function iterates over an array of OSSL_PARAM structures and prints the
+ * value of each parameter to the specified BIO stream. The parameter type is
+ * detected, and the value is printed accordingly.
+ *
+ * @param p Array of OSSL_PARAM structures to be printed.
+ * @param b BIO stream to print the parameter values to.
+ *
+ * The supported parameter types are:
+ * - OSSL_PARAM_UNSIGNED_INTEGER: Unsigned integer values.
+ * - OSSL_PARAM_INTEGER: Signed integer values.
+ * - OSSL_PARAM_UTF8_PTR: Pointer to a UTF-8 string.
+ * - OSSL_PARAM_UTF8_STRING: UTF-8 string.
+ * - OSSL_PARAM_OCTET_PTR/OCTET_STRING: Binary data (size in bytes).
+ */
+#ifndef FIPS_MODULE
+static void trace_print_param_values(const OSSL_PARAM p[], BIO *b)
+{
+    int64_t i;
+    uint64_t u;
+# ifndef OPENSSL_SYS_UEFI
+    double d;
+# endif
+
+    for (; p->key != NULL; p++) {
+        BIO_printf(b, "%s: ", p->key);
+        switch (p->data_type) {
+        case OSSL_PARAM_UNSIGNED_INTEGER:
+            if (OSSL_PARAM_get_uint64(p, &u))
+                BIO_printf(b, "%llu\n", (unsigned long long int)u);
+            else
+                BIO_printf(b, "error getting value\n");
+            break;
+        case OSSL_PARAM_INTEGER:
+            if (OSSL_PARAM_get_int64(p, &i))
+                BIO_printf(b, "%lld\n", (long long int)i);
+            else
+                BIO_printf(b, "error getting value\n");
+            break;
+        case OSSL_PARAM_UTF8_PTR:
+            BIO_printf(b, "'%s'\n", *(char **)(p->data));
+            break;
+        case OSSL_PARAM_UTF8_STRING:
+            BIO_printf(b, "'%s'\n", (char *)p->data);
+            break;
+        case OSSL_PARAM_OCTET_PTR:
+        case OSSL_PARAM_OCTET_STRING:
+            BIO_printf(b, "<%zu bytes>\n", p->data_size);
+            break;
+# ifndef OPENSSL_SYS_UEFI
```

Comment:
I wouldn't handle this around this way - the type is known - someone can elect to pass it in - the platform doesn't have the conversion support routines - but it is a known type - handled this way it is seen as unknown - and it isn't - it is unsupported on this platform. Look at the other usage where OSSL_PARAM_REAL is on the "outside" and the ifdef check is on the "inside".